{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aceaefc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removing triallelic SNPs \n",
    "# (If a triallelic SNP is significant, keep it and remove the other one. If neither are significant then remove all)\n",
    "\n",
    "import pandas as pd \n",
    "\n",
    "# import the GWAS imputed data\n",
    "\n",
    "mycols = ['chrom','position', 'ref_allele', 'effect_allele', 'minor_allele', 'minor_AF', 'expected_case_minor_AC', 'low_confidence_variant', 'n_complete_samples', 'AC', 'ytx', 'beta', 'se', 'tstat', 'pval']\n",
    "\n",
    "df_imputed = {}\n",
    "\n",
    "for i in range(1, 23):\n",
    "    df_imputed[i] = pd.read_csv(f\"{i}_fullGWAS.tsv\", sep=r\"[\\t:]\", names=mycols, engine='python')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "42b4db30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   chrom  position ref_allele effect_allele minor_allele  minor_AF  \\\n",
      "0      1     69487          G             A            A  0.000006   \n",
      "1      1     69569          T             C            C  0.000188   \n",
      "2      1    139853          C             T            T  0.000006   \n",
      "3      1    693731          A             G            G  0.115824   \n",
      "4      1    707522          G             C            C  0.097281   \n",
      "\n",
      "   expected_case_minor_AC  low_confidence_variant  n_complete_samples  \\\n",
      "0                 1.11928                    True              360420   \n",
      "1                36.54850                    True              360420   \n",
      "2                 1.10237                    True              360420   \n",
      "3             22502.00000                   False              360420   \n",
      "4             18899.50000                   False              360420   \n",
      "\n",
      "            AC          ytx      beta        se     tstat      pval  \n",
      "0      4.15294      1.01176 -0.089465  0.215551 -0.415051  0.678105  \n",
      "1    135.60800     29.21180 -0.051813  0.038658 -1.340300  0.180150  \n",
      "2      4.09020      1.00000 -0.087803  0.215557 -0.407328  0.683767  \n",
      "3  83490.20000  22499.40000 -0.000223  0.001689 -0.131812  0.895133  \n",
      "4  70123.70000  18825.90000 -0.001348  0.001899 -0.710169  0.477600  \n",
      "dict_keys([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22])\n"
     ]
    }
   ],
   "source": [
    "print(df_imputed[1].head()) # check to see it's there\n",
    "print(df_imputed.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "594a1d71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now find the duplicated columns \n",
    "\n",
    "duplicates_df = {}\n",
    "\n",
    "for i in range(1,23): \n",
    "    duplicates_df[i] = df_imputed[i][df_imputed[i].duplicated(subset=['position'],keep=False)]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "69c94bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# put the duplicates in an array\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "array = {}\n",
    "\n",
    "for i in range(1,23):\n",
    "    array[i] = duplicates_df[i][[\"position\"]].to_numpy() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5f8725f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1186665])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "array[1][3] # array[i] is the duplicates in chromosome i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "15d2788c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>expected_case_minor_AC</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>634</th>\n",
       "      <td>1</td>\n",
       "      <td>901922</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000357</td>\n",
       "      <td>69.3343</td>\n",
       "      <td>True</td>\n",
       "      <td>360420</td>\n",
       "      <td>257.255</td>\n",
       "      <td>67.851</td>\n",
       "      <td>0.001520</td>\n",
       "      <td>0.027620</td>\n",
       "      <td>0.055029</td>\n",
       "      <td>0.956115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>635</th>\n",
       "      <td>1</td>\n",
       "      <td>901922</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.004481</td>\n",
       "      <td>870.6390</td>\n",
       "      <td>False</td>\n",
       "      <td>360420</td>\n",
       "      <td>3230.380</td>\n",
       "      <td>879.949</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.007620</td>\n",
       "      <td>0.034551</td>\n",
       "      <td>0.972438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2363</th>\n",
       "      <td>1</td>\n",
       "      <td>1186665</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.123826</td>\n",
       "      <td>24056.7000</td>\n",
       "      <td>False</td>\n",
       "      <td>360420</td>\n",
       "      <td>89259.000</td>\n",
       "      <td>24129.000</td>\n",
       "      <td>0.001031</td>\n",
       "      <td>0.001545</td>\n",
       "      <td>0.666929</td>\n",
       "      <td>0.504818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2364</th>\n",
       "      <td>1</td>\n",
       "      <td>1186665</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.010815</td>\n",
       "      <td>2101.1700</td>\n",
       "      <td>False</td>\n",
       "      <td>360420</td>\n",
       "      <td>7796.090</td>\n",
       "      <td>2114.470</td>\n",
       "      <td>0.004348</td>\n",
       "      <td>0.005223</td>\n",
       "      <td>0.832427</td>\n",
       "      <td>0.405168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5091</th>\n",
       "      <td>1</td>\n",
       "      <td>1885615</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001063</td>\n",
       "      <td>206.4640</td>\n",
       "      <td>False</td>\n",
       "      <td>360420</td>\n",
       "      <td>766.055</td>\n",
       "      <td>207.090</td>\n",
       "      <td>0.003304</td>\n",
       "      <td>0.016612</td>\n",
       "      <td>0.198905</td>\n",
       "      <td>0.842337</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  position ref_allele effect_allele minor_allele  minor_AF  \\\n",
       "634       1    901922          G             A            A  0.000357   \n",
       "635       1    901922          G             C            C  0.004481   \n",
       "2363      1   1186665          G             A            A  0.123826   \n",
       "2364      1   1186665          G             T            T  0.010815   \n",
       "5091      1   1885615          A             C            C  0.001063   \n",
       "\n",
       "      expected_case_minor_AC  low_confidence_variant  n_complete_samples  \\\n",
       "634                  69.3343                    True              360420   \n",
       "635                 870.6390                   False              360420   \n",
       "2363              24056.7000                   False              360420   \n",
       "2364               2101.1700                   False              360420   \n",
       "5091                206.4640                   False              360420   \n",
       "\n",
       "             AC        ytx      beta        se     tstat      pval  \n",
       "634     257.255     67.851  0.001520  0.027620  0.055029  0.956115  \n",
       "635    3230.380    879.949  0.000263  0.007620  0.034551  0.972438  \n",
       "2363  89259.000  24129.000  0.001031  0.001545  0.666929  0.504818  \n",
       "2364   7796.090   2114.470  0.004348  0.005223  0.832427  0.405168  \n",
       "5091    766.055    207.090  0.003304  0.016612  0.198905  0.842337  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicates_df[1].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "df46dbd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_df = {}\n",
    "\n",
    "for i in range(1,23):\n",
    "    cleaned_df[i] = df_imputed[i][df_imputed[i].apply(lambda x: (x[\"position\"] in array[i] and x[\"pval\"] < 0.00001) or x[\"position\"] not in array[i], axis=1)]\n",
    "                          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "84c8d3ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "        chrom  position ref_allele effect_allele minor_allele  minor_AF  \\\n",
      "218016     15  75139426          A             G            A  0.415469   \n",
      "218017     15  75139426          A             T            A  0.223686   \n",
      "\n",
      "        expected_case_minor_AC  low_confidence_variant  n_complete_samples  \\\n",
      "218016                 80716.5                   False              360420   \n",
      "218017                 43457.2                   False              360420   \n",
      "\n",
      "              AC       ytx      beta        se    tstat          pval  \n",
      "218016  421353.0  112341.0 -0.007275  0.001036 -7.02231  2.186140e-12  \n",
      "218017  559598.0  149737.0 -0.009563  0.001262 -7.57631  3.563640e-14  \n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [chrom, position, ref_allele, effect_allele, minor_allele, minor_AF, expected_case_minor_AC, low_confidence_variant, n_complete_samples, AC, ytx, beta, se, tstat, pval]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "for i in range(1,23):\n",
    "    print(cleaned_df[i][cleaned_df[i].duplicated(subset=[\"position\"],keep=False)].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "05c630e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(948317, 15)\n",
      "(1030243, 15)\n",
      "(870283, 15)\n",
      "(878477, 15)\n",
      "(792479, 15)\n",
      "(807681, 15)\n",
      "(711559, 15)\n",
      "(679865, 15)\n",
      "(530103, 15)\n",
      "(615661, 15)\n",
      "(608395, 15)\n",
      "(572274, 15)\n",
      "(432288, 15)\n",
      "(389047, 15)\n",
      "(342441, 15)\n",
      "(382762, 15)\n",
      "(327510, 15)\n",
      "(339309, 15)\n",
      "(278584, 15)\n",
      "(269395, 15)\n",
      "(161590, 15)\n",
      "(162893, 15)\n",
      "2\n",
      "4\n",
      "1\n",
      "2\n",
      "0\n",
      "6\n",
      "0\n",
      "4\n",
      "0\n",
      "5\n",
      "7\n",
      "3\n",
      "2\n",
      "2\n",
      "3\n",
      "1\n",
      "1\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "# see how many duplicates there are \n",
    "\n",
    "for i in range(1,23):\n",
    "    print(cleaned_df[i].shape)\n",
    "    \n",
    "for i in range(1,23):\n",
    "    print(cleaned_df[i].shape[0] - (df_imputed[i].shape[0] - duplicates_df[i].shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2b7b5d4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save unqiue output as a tsv file \n",
    "\n",
    "for i in range(1,23):\n",
    "    cleaned_df[i].to_csv(f\"{i}allcleaned.tsv\", sep=\"\\t\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5a34332",
   "metadata": {},
   "outputs": [],
   "source": [
    "### make cleaned file into a VEP input file\n",
    "\n",
    "for X in {1..22}\n",
    "do\n",
    "awk 'BEGIN{FS=\"\\t\"} {print $2\"\\t\"$3\"\\t\"$3\"\\t\"$4\"/\"$5\"\\t\"\"+\"}' ${X}allcleaned.tsv >> ${X}_VEP1.tsv\n",
    "done\n",
    " "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
